sm/agent/src/state.rs

```diff
@@ -0,0 +1,239 @@
+use crate::error::StateError;
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use std::{
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+use tedge_config::TEdgeConfigLocation;
+
+use tokio::{fs, io::AsyncWriteExt};
+
+#[derive(Debug)]
+pub struct AgentStateRepository {
+    state_repo_path: PathBuf,
+    state_repo_root: PathBuf,
+}
+
+#[async_trait]
+pub trait StateRepository<T> {
+    type Error;
+    async fn load(&self) -> Result<T, Self::Error>;
+    async fn store(&self, state: &T) -> Result<(), Self::Error>;
+    async fn clear(&self) -> Result<T, Self::Error>;
+}
+
+#[async_trait]
+impl StateRepository<State> for AgentStateRepository {
+    type Error = StateError;
+
+    async fn load(&self) -> Result<State, StateError> {
+        match fs::read(&self.state_repo_path).await {
+            Ok(bytes) => Ok(toml::from_slice::<State>(bytes.as_slice())?),
+
+            Err(err) => Err(StateError::IOError(err)),
+        }
+    }
+
+    async fn store(&self, state: &State) -> Result<(), StateError> {
+        let toml = toml::to_string_pretty(&state)?;
+
+        // Create `$HOME/.tedge` or `/etc/tedge` directory in case it does not exist yet
+        if !self.state_repo_root.exists() {
+            let () = fs::create_dir(&self.state_repo_root).await?;
+        }
+
+        let mut temppath = self.state_repo_path.clone();
+        temppath.set_extension("tmp");
+
+        let () = atomically_write_file(temppath, &self.state_repo_path, toml.as_bytes()).await?;
+
+        Ok(())
+    }
+
+    async fn clear(&self) -> Result<State, Self::Error> {
+        let state = State {
+            operation_id: None,
+            operation: None,
+        };
+        let () = self.store(&state).await?;
+
+        Ok(state)
+    }
+}
+
+impl AgentStateRepository {
+    pub fn new(config_location: &TEdgeConfigLocation) -> Self {
+        let mut state_repo_root = config_location.tedge_config_root_path.clone();
+        state_repo_root.push(PathBuf::from_str(".agent").expect("infallible"));
+
+        let mut state_repo_path = state_repo_root.clone();
+        state_repo_path.push(PathBuf::from_str("software").expect("infallible"));
```

Comment:
`software` is a bit vague. I prefer `operation-state` or better `current-operation`.